### PR TITLE
UCT/TCP: Fix connect error handling

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Just return an error in case of  `ucs_socket_connect` failure 

## Why ?

We have to not move EP from `CLOSED` connection state to `CLOSED` state (there is an assertion for that)

## How ?

Simplify `uct_tcp_cm_conn_start`

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/12363/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 25-Sep-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/cb187f444682fa4629a17b80c95e21bd2696ffca/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/250c53722619c62b57cd60e09429345432083cfa/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/30869b46a250d8297a65b57c91f92f71d09f84fc/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/6ac4854d38985f3b5049ff30ed8b2c37ddc1c1a6/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/478a640c4af52b82e92a0c01037f46f8c78ab294/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/d08cb10bbeba5749b7c94e3010c723eb5ace17ae/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/241317524fec8797782d8a2682190265e9013d42/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/3ab60bbcee39b4681ecb27d3f45dafc305b2d7b0/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/cb88e60a974a1639f1feadbd78f77b74bd115213/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/106d0bf29434aefa2c175070051c4145c5544af8/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/682812020d7d277c13e0ba88959b36a6d3f47c70/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/3f08a693316811e2b0e23d15525e08f220e85df8/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/2d391c391e4b6e9b53ffb96bc781e454550de9fb/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/c4ee6cf38fb63f602106330e3292477b613ab18b/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/1db9b8a67915bac609f5b58237c26c6c33fef9c1/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/5a666ace30eecceb004ec7ba760372c777563ed7/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/94981f97236ff81885be2d65bd53194e86d40294/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/ca83c72ed8d819085f001cee1371b08d8f1330a6/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/d1ef083404926172b21b708b3dd84a58606547e3/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/5a0f9d9322f6be75361153d3294290acc414f31d/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/33a5bc19fdb7fad713e2f2e4f1599f8d68a92f88/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/194bcb478f4be3ac462d10d9c65014da71312c9d/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/c5c545689e1d749647cdddcbf169fa1a081ea325/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/e8fae58636e8eadeb0b80e28de92e850dc0f71d6/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/afeedaf90625262b13f6f37bb0982308537f4069/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/b60fd561e898fe94ae850dc20127cab924c91895/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/9462a17714f30b6345ec97f0e7b9d2b41440a1e9/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/13fdfcb54090d4e3b626d5ea7cd478a112fba6b6/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e31c3f6bf4daf030bd4621416b8a1266/raw/0db0651a85b31bfb08e5259d3cc09e9895dd3e95/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

File: src/uct/tcp/tcp_cm.c [hunk 1]
```diff
@@ -545,31 +545,25 @@ unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
...
-    uct_tcp_ep_conn_state_t new_conn_state;
-    uint32_t req_events;
     ucs_status_t status;
 
     status = ucs_socket_connect(ep->fd, (const struct sockaddr*)&ep->peer_addr);
-    if (status == UCS_INPROGRESS) {
-        new_conn_state  = UCT_TCP_EP_CONN_STATE_CONNECTING;
-        req_events      = UCS_EVENT_SET_EVWRITE;
-        status          = UCS_OK;
+    if (UCS_PTR_IS_ERR(status)) {
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_cm.c]: status is not a pointer, need to use `UCS_STATUS_IS_ERR`

File: src/uct/tcp/tcp_cm.c [hunk 2]
```diff
@@ -545,31 +545,25 @@ unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
...
-        new_conn_state  = UCT_TCP_EP_CONN_STATE_CLOSED;
-        req_events      = 0;
+        uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAITING_ACK);
+        uct_tcp_ep_mod_events(ep, UCS_EVENT_SET_EVREAD, 0);
     }
 
-    uct_tcp_cm_change_conn_state(ep, new_conn_state);
-    uct_tcp_ep_mod_events(ep, req_events, 0);
-    return status;
+    return UCS_OK;
```
Comment (evgeny-leksikov) [inline src/uct/tcp/tcp_cm.c]: seems like this code is unreachable, then last check `} else if (status == UCS_OK) {` is not needed, it's always true

File: src/uct/tcp/tcp_cm.c (hunk 1 above)
Comment (dmitrygx) [reply src/uct/tcp/tcp_cm.c]: done

File: src/uct/tcp/tcp_cm.c (hunk 2 above)
Comment (dmitrygx) [reply src/uct/tcp/tcp_cm.c]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/12364/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 25-Sep-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/f4bc20087e6776fd4b8f8a45c7ffa2eb85581ead/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/333668b28c91a1a9fb49625212dfc680095220b3/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/d8eadc24ae2ee3ae3618c679784b815d7579703c/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/fe5f10286964dd1e764ca26addfd252b91ff0772/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/c61b29417356803570fa2e21aa046571799a0c8d/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/5f08c104215595175d64db5cebfe8a35a53d53f2/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/50206a28dbea403c0b64dbc3f735d7529ae0fbd4/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/27a1c52b106624d2e8bcf929c9e5394ffa63a9f6/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/2e62f177259749c5da61f2eeb557a135fe932a25/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/6626fe9a9b0c61268e62c438efb1088b98f7f669/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/f726668d539d90d86e64993590a5dca0c85f05ba/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/d17dee49802edbc79f243e8b3fc087a2748313bb/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/8dc112e642f68a5ff57c4c2b9d75b91f7bfbf1a2/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/c332c3b2f15d7c9d4d09f7acb9afb11d2dea2efd/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/b000b1048edcabc7a754b0a311f3840b0dcf4bca/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/fe3d7f7f3e97d3f6390284028b71cad2e91f34dd/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/71f539307f806e80b9758bf27ec3e161016995be/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/6a18aa62aacb65b4f4330e846c558d3e483307d5/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/769248c7003a7e940df1d809289f7e9463a59a60/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/fe64a8d99aa028844af6399f241166c07445d5ac/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/c8d79aa2ac257ac757ebf3e2f47f5469eb29ba03/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/0b1fda987d969de9819d8730224349c35b975423/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/292735cccba6022da35ad2e13f184ebe0696101b/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/9364a66e4a307ab9b0e22fe19f9854b6fa5dc0da/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/cad95a3ecadd68eeae7a5edc71b0693426a3d366/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/00f6b7dc8bfcc13fa9310d048a49283fd70f1fc0/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/35035c6c6b40c86f82f88099553401a438caee03/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/66e29c25e9d08235ca9bf7f7830e3a9422806943/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/90c3e4c0ace770366059fa32c1992563/raw/e83d1fe31b6ba8ca2fa4d815eed0b49331788f88/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by brminich** [comment]
what scenario does it fix?

**Comment by dmitrygx** [comment]
> what scenario does it fix?

Fixes assertion when `ucs_socket_connect()` fails (i.e. return value < 0), TCP moves EP from `CLOSED` to `CLOSED` state - this is a bad and shouldn't be done

**Comment by brminich** [comment]
ic, because it is already in `closed` state

**Comment by dmitrygx** [comment]
> ic, because it is already in `closed` state

yes

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

**Pull request merged**